publish pfSense implementation under Xen - boot optimization (#175)
3mdeb/news-and-ideas: blog/content/post/2019-12-13-pfsense-boot-under-xen.md

---
title: pfSense firewall boot process optimization under Xen hypervisor
abstract: In previous article we introduce our implementation of pfSense under
          Xen. Now, we want to show how you can improve boot process and reduce
          virutalized pfSense boot time to minimum.
cover: /covers/pfsense-logo.png
author: piotr.kleinschmidt
layout: post
published: true
date: 2019-12-13
archives: "2019"

tags:
  - pfSense
  - Xen
  - virtualization
  - hypervisor
  - firewall
categories:
  - Firmware
  - OS dev

---

## Introduction

That article is continuation of [pfSense firewall implementation under Xen
hypervisor](https://blog.3mdeb.com/2019/2019-11-06-pfsense-under-xen-introduction/).
If you haven't read that yet, I strongly recommend to get familiar with it

> before that article. There are basic information about virtualization,

Reviewer: ```suggestion
before reading this article. There is basic information about virtualization,
```

hypervisor and our build configuration.

Here, I focused only on **pfSense booting optimization**. You will find out what
problems occurred and how we handled them. Also, you will see that meticulous

> analysis is a key in solving any problems and very often it pays to devote time

Reviewer: ```suggestion
analysis is a key in solving any problems and very often it pays off to devote time
```

for it.

## pfSense booting

As I mentioned in previous article, we made actually 2 implementations of
pfSense: bare-metal and with Xen. First one is used as reference one. If
virtualized pfSense could approach to reference values, then it is considered as
well-implemented.

What we wanted to gain is to have boot time as short as possible. For bare-metal
case boot time is considered as time from power on platform to enter pfSense
main menu. For VM implementation, it is measured from creating VM in Xen to
enter main menu. In both cases every possible delay related to firmware was
turned off.

### Bare-metal pfSense

First, we needed to know how long bare-metal pfSense boots. As I mentioned, it
will be our reference sample to every next test.

Every delay which could occur in coreboot or SeaBIOS is turned off. Also
configuration for pfSense is optimized. We disabled `pfSense boot menu` (by the
way determined by vendors as `beastie menu`). Also we disabled auto-boot time,

> so the kernel is loaded automatically without any delay. Follow logs show what

Reviewer: ```suggestion
so the kernel is loaded automatically without any delay. Following logs show what
```

maybe we should look for the source of the problem somewhere else...

### Debug Xen

If the problem doesn't lie in pfSense itself, then it certainly must lie higher -
in the hypervisor. Fortuantely, Xen has a very helpful tool for debugging
purpose. We used `Xentrace` and `Xenalyze`. Usage of them is well-descripted
[here](https://xenproject.org/2012/09/27/tracing-with-xentrace-and-xenalyze/),
so I won't duplicate that article. What you might need to know, `Xentrace` is a
tool for monitoring Xen performance. It gives logs about almost every parameter
which is useful from debugging point of view. For example, you can see what CPUs
are currently used by what VM. What is a state of VM and what interrupts change
that state. How long it was running and how long it was in idle state and many,
many other. `Xenalyze` transpose those logs to human-friendly form. It is then
easier to analyze them and search for potential issues.

Our priority is to find out why there is 2-minute delay in boot process and how
we can reduce it. To do this we collected Xentrace logs from that particular
stage of booting (actually we took only 12 seconds of that stage, because

> Xentrace logs are very large and we didn't want to get out of free space). After

Reviewer: ```suggestion
Xentrace logs are very large and we didn't want to run out of free space). After
```

> using Xenalyze on those logs, we achieved interesting output. **pfSense VM

Reviewer: ```suggestion
using `Xenalyze` on those logs, we achieved interesting output. **pfSense VM
```

during boot process is most of the time in offline state caused by VM_EXIT_IOIO
interrupt**. That interrupt is triggered every time when software tries to
read/write data to disk. It gives a clue that the problem might be caused by the
way of how SD card (on which pfSense image is) is emulated by Xen.

Every VM creation can be configured by a `.cfg` file. That file is passed to
Xen's `create` command as an argument. If you want to change anything about your

> VM at the creation stage, that file is place when you add changes. For our issue

Reviewer: ```suggestion
VM at the creation stage, that file is the place where you add changes. For our issue
```

we made one significant modification. So far, SD card with pfSense image was
emulated as `hd[X]` device. It means that pfSense VM has communicated with it
via **emulated IDE controller**. We decided to change it to `sd[X]`, which means
**emulated SCSI controller**. Theoretically, both IDE and SCSI operations speed
are almost the same. The difference is in the number of cycles to access the
registers during R/W operations. SCSI needs less of them than IDE. Less

> registers access = less IOIO interrupts = less exits to hypervisor.

Reviewer: ```suggestion
register accesses = less IOIO interrupts = less exits to hypervisor.
```

00:01:40  LAN (lan)       -> xn1        -> v4: 192.168.1.1/24
00:01:40
00:01:40  0) Logout (SSH only)                  9) pfTop
00:01:40  1) Assign Interfaces                 10) Filter Logs
00:01:40  2) Set interface(s) IP address       11) Restart webConfigurator
00:01:40  3) Reset webConfigurator password    12) PHP shell + pfSense tools
00:01:40  4) Reset to factory defaults         13) Update from console
00:01:40  5) Reboot system                     14) Enable Secure Shell (sshd)
00:01:40  6) Halt system                       15) Restore recent configuration
00:01:40  7) Ping host                         16) Restart PHP-FPM
00:01:40  8) Shell
00:01:40
```

>Timestamps are given as stopwatch format counting from 00:00:00. Therefore,
last value is boot time

Attempt 3: **pfSense in VM boot time** : **00:01:40**

There is no more unnecessary delays in booting. Also the entire process is only
19 seconds longer than bare-metal comparative one. We should consider it as a
success and it can be said that we gained our set goal. After applying all above
changes, booting pfSense under Xen hypervisor can be now described as optimized.

## Summary

> As you probably have often found out, even well-described processes could bring

Reviewer: ```suggestion
As you probably have found out, even well-described processes could often bring
```

unexpected issues. Especially nowadays, when performance and reliability are

> significant features of every hardware and software. Above blogpost proves that

Reviewer: ```suggestion
significant features of every hardware and software. Above blog post proves that
```
